internal/controller/apiextensions/definition/reconciler.go

```diff
@@ -471,19 +470,27 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reco
 	u.SetGroupVersionKind(xrGVK)
 
 	name := composite.ControllerName(d.GetName())
-	var ca cache.Cache
+	var xrCache cache.Cache // late initialized below
 	watches := []controller.Watch{
 		controller.For(u, &handler.EnqueueRequestForObject{}),
 		// enqueue composites whenever a matching CompositionRevision is created
-		controller.TriggeredBy(source.Kind(r.mgr.GetCache(), &v1.CompositionRevision{}), handler.Funcs{
-			CreateFunc: composite.EnqueueForCompositionRevisionFunc(ck, r.mgr.GetCache().List, r.log),
+		//
+		// Note: we explicitly use the xrCache for composition revisions to sync
+		// the life-cycle of the source and the composite list, and to avoid
+		// leaking these triggers (source.Kind does not remove a handler ever).
+		controller.TriggeredBy(&lazySyncingSource{Create: func() source.SyncingSource {
+			return source.Kind(xrCache, &v1.CompositionRevision{})
+		}}, handler.Funcs{
+			CreateFunc: composite.EnqueueForCompositionRevisionFunc(ck, func(ctx context.Context, list client.ObjectList, opts ...client.ListOption) error {
+				return xrCache.List(ctx, list, opts...)
+			}, r.log),
```

Comment:
Style nit: This is a bit of a handful to read due to the nested function definitions with long signatures.

Reply:
have made it similar to the other source. A little bit easier to read. The closured for delaying evaluation are little ugly. But Go doesn't allow a better syntax. Wished there was a `foo(bla, lazy xrCache, ...)`.